File: dotnet/src/PuzzleSolver.NumbersGame/Number.cs
namespace PuzzleSolver.NumbersGame;

public class Number
{
    public int Value { get; init; }
    
    public NumberCategory Category { get; init; }

    private Number(int value) : this(value, NumberCategory.Custom)
    {
    } 
    
    private Number(int value, NumberCategory category)
    { 
        Value = value;
        Category = category;
    }

    public override string ToString()
    {
        return $"{Value} ({Category})";
    }

    internal static Number Generate(int number)
    {
        if (_commonNumbers.TryGetValue(number, out var commonNumber))
        {
            return commonNumber;
        }
        
        return new Number(number);
    }

    private static readonly IReadOnlyDictionary<int, Number> _commonNumbers = new Dictionary<int, Number>(14)
    {
        { 1, new Number(1, NumberCategory.Small) },
        { 2, new Number(2, NumberCategory.Small) },
        { 3, new Number(3, NumberCategory.Small) },
        { 4, new Number(4, NumberCategory.Small) },
        { 5, new Number(5, NumberCategory.Small) },
        { 6, new Number(6, NumberCategory.Small) },
        { 7, new Number(7, NumberCategory.Small) },
        { 8, new Number(8, NumberCategory.Small) },
        { 9, new Number(9, NumberCategory.Small) },
        { 10, new Number(10, NumberCategory.Small) },
        { 25, new Number(25, NumberCategory.Large) },
        { 50, new Number(50, NumberCategory.Large) },
        { 75, new Number(75, NumberCategory.Large) },
        { 100, new Number(100, NumberCategory.Large) }
    }.AsReadOnly();
}

public enum NumberCategory
{
    Small,
    Large,
    Custom
}

File: python/number.py
from dataclasses import dataclass
from enum import Enum
from typing import Self

class NumberCategory(Enum):
    SMALL = "small"
    LARGE = "large"
    CUSTOM = "custom"

class Number:
    def __init__(self, value: int) -> None:
        """
        Initializes a Number with a value and determines its category.
        
        :param value: The integer value of the number.
        """
        self._value = value

        if 1 <= value <= 10:
            self._category = NumberCategory.SMALL
        elif value in [25, 50, 75, 100]:
            self._category = NumberCategory.LARGE
        else:
            self._category = NumberCategory.CUSTOM

    @property
    def value(self) -> int:
        return self._value
    
    @property
    def category(self) -> NumberCategory:
        return self._category
    
    def __eq__(self, other: object) -> bool:
        """
        Checks equality with another Number.
        
        :param other: Another Number instance.
        :return: True if both numbers should be considered equal, False otherwise.
        """

        if not isinstance(other, Number):
            return False

        return self.value == other.value
    
    def __hash__(self) -> int:
        """
        Returns a hash of the number.
        
        :return: The hash value of the number.
        """
        return hash(self.value)
    
    def __str__(self) -> str:
        """
        Returns a string representation of the number.
        
        :return: The string representation of the number.
        """
        return f"{self.value} ({self.category.value.title()})"
